rest_server/IOMgrConnection.go

```diff
@@ -0,0 +1,393 @@
+		databaseEngineRestRequest.ObjectId = objectId
+	}
+	siodbLoggerPool.Debug("databaseEngineRestRequest: %v", databaseEngineRestRequest)
+	siodbLoggerPool.Debug("IOMgrConn.RequestID: %v", requestId)
+	siodbLoggerPool.Debug("IOMgrConn.RequestID++: %v", IOMgrConn.RequestID)
+
+	if _, err := IOMgrConn.writeMessage(DATABASEENGINERESTREQUEST, &databaseEngineRestRequest); err != nil {
+		return requestId, fmt.Errorf("Unable to write message to IOMgr: %v", err)
+	}
+
+	return requestId, nil
+}
+
+func (IOMgrConn *IOMgrConnection) readIOMgrResponse(requestID uint64) (err error) {
+
+	var databaseEngineResponse SiodbIomgrProtocol.DatabaseEngineResponse
+
+	if _, err := IOMgrConn.readMessage(DATABASEENGINERESPONSE, &databaseEngineResponse); err != nil {
+		return fmt.Errorf("Unable to read response from IOMgr: %v", err)
+	}
+	siodbLoggerPool.Debug("databaseEngineResponse: %v", databaseEngineResponse)
+	siodbLoggerPool.Debug("databaseEngineResponse.RequestId: %v", databaseEngineResponse.RequestId)
+	siodbLoggerPool.Debug("IOMgrConn.RequestID: %v", IOMgrConn.RequestID)
+
+	if databaseEngineResponse.RequestId != requestID {
+		return fmt.Errorf("request IDs mismatch")
+	}
+
+	if len(databaseEngineResponse.Message) > 0 {
+		return fmt.Errorf("code: %v, message: %v", databaseEngineResponse.Message[0].GetStatusCode(), databaseEngineResponse.Message[0].GetText())
+	}
+
+	return nil
+}
+
+func (IOMgrConn *IOMgrConnection) readChunkedJSON(c *gin.Context) (err error) {
+
+	// Stream through HTTP > 1.1
+	// https://stackoverflow.com/questions/29486086/how-http2-http1-1-proxy-handle-the-transfer-encoding
+
+	var IOMgrChunkSize uint32
+	var IOMgrbuffSize uint32
+	var IOMgrBytesRead int
+	var IOMgrbuffRemain []byte
+	var JSONChunk string
+	var JSONChunkBuff string
+
+	for {
+
+		// Get Current IOMgr chunk Size
+		if _, IOMgrChunkSize, err = IOMgrConn.readVarint32(); err != nil {
+			return fmt.Errorf("Not able to read chunk size from IOMgr: %v", err)
+		}
+		siodbLoggerPool.Debug("IOMgrChunkSize: %v", IOMgrChunkSize)
+
+		if IOMgrChunkSize == 0 {
+			break
+		}
+
+		for IOMgrChunkSize != 0 {
+			if IOMgrChunkSize < IOMgrChunkMaxBufferedSize {
+				IOMgrbuffSize = IOMgrChunkSize
+				IOMgrChunkSize = 0
+			} else if IOMgrChunkSize >= IOMgrChunkMaxBufferedSize {
+				IOMgrbuffSize = IOMgrChunkMaxBufferedSize
+				IOMgrChunkSize = IOMgrChunkSize - IOMgrChunkMaxBufferedSize
+			}
+			siodbLoggerPool.Debug("IOMgrbuffSize: %v", IOMgrbuffSize)
+
+			IOMgrbuff := make([]byte, uint32(len(IOMgrbuffRemain))+IOMgrbuffSize)
+
+			// Form JSONChunk
+			IOMgrBytesRead, err = io.ReadFull(IOMgrConn, IOMgrbuff[len(IOMgrbuffRemain):])
+			siodbLoggerPool.Debug("Bytes read from IOMgr: %v", IOMgrBytesRead)
+			if len(IOMgrbuffRemain) > 0 {
+				bc := copy(IOMgrbuff, IOMgrbuffRemain)
+				siodbLoggerPool.Debug("Bytes copied from IOMgrbuffRemain into IOMgrbuff : %v", bc)
+				IOMgrbuffRemain = nil
+			}
+
+			for len(IOMgrbuff) > 0 {
+				UTF8rune, size := utf8.DecodeRune(IOMgrbuff)
```

Comment:
why at all this checking of truncated utf8 is neede? isn't it responsibility of client to gather multibyte characters? I suggest to make a test. I feel this can be dropped away and you just can pass bytes as is from iomgr to http client.

Reply:
@cxxman yes you're right. I was wrongly no using chunked encoding. Instead, I was using no encoding and sending chunks as is which curl interpreted as separated pieces of data. That why I had those broken UTF-8 points. It is now fixed for the coming commits.